src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -72,50 +73,75 @@ internal int FindMetricAggregators(ReadOnlySpan<KeyValuePair<string, object>> ta
-                    seqKey = new string[len];
-                    tagKey.CopyTo(seqKey, 0);
-
-                    value2metrics = new Dictionary<object[], int>(new ObjectArrayEqualityComparer());
-                    this.keyValue2MetricAggs.Add(seqKey, value2metrics);
+                    // Some other thread added value2metrics, take it.
+                    // The one created by this thread is wasted.
+                    this.keyValue2MetricAggs.TryGetValue(tagKey, out value2metrics);
                 }
+            }
 
-                // GetOrAdd by TagValue at 2st Level of 2-level dictionary structure.
-                // Get back Metrics[].
-                if (!value2metrics.TryGetValue(tagValue, out aggregatorIndex))
+            // GetOrAdd by TagValue at 2st Level of 2-level dictionary structure.
+            // Get back Metrics[].
+            if (!value2metrics.TryGetValue(tagValue, out aggregatorIndex))
+            {
+                this.metricPointIndex++;
+                aggregatorIndex = this.metricPointIndex;
+                if (aggregatorIndex >= MaxMetricPoints)
                 {
-                    this.metricPointIndex++;
-                    aggregatorIndex = this.metricPointIndex;
-                    if (aggregatorIndex >= MaxMetricPoints)
-                    {
-                        // sorry! out of data points.
-                        return -1;
-                    }
+                    // sorry! out of data points.
+                    // TODO: This doesn't account for "wasted space"
+                    // in the array. Need to reclaim them when we
+                    // really need. i.e inside this if block.
+                    // this can take explicit lock, as this
+                    // condition is not the common code path.
+                    return -1;
+                }
 
-                    // Note: We are using storage from ThreadStatic, so need to make a deep copy for Dictionary storage.
+                // Note: We are using storage from ThreadStatic, so need to make a deep copy for Dictionary storage.
+
+                if (seqKey == null)
+                {
+                    seqKey = new string[len];
+                    tagKey.CopyTo(seqKey, 0);
+                }
 
-                    if (seqKey == null)
+                var seqVal = new object[len];
+                tagValue.CopyTo(seqVal, 0);
+                if (!value2metrics.TryAdd(seqVal, aggregatorIndex))
+                {
+                    // This thread lost the Add.
+                    // Some other thread added aggregatorIndex, take it.
+                    // aggregatorIndex incremented by this thread is
+                    // now a "wasted space" in the array.
+                    value2metrics.TryGetValue(tagValue, out aggregatorIndex);
+                    ref var metricPoint = ref this.metrics[aggregatorIndex];
+                    while (metricPoint.StartTime == default)
```

Comment:
might be able to find alternate here...

Reply:
How about using `SpinWait.SpinUntil (() => { return metricPoint.StartTime != default });`?

Reply:
yes. 
Other issue I actually meant here : checking StartTime is not safe either as its not atomically readable. Will use some other field to indicate if MetricPoint is initialized.